src/app/stream-deck/msg-handlers.ts

```diff
@@ -0,0 +1,311 @@
+    const allItems = allItemsSelector(state);
+    const loadout = randomLoadout(
+      store,
+      allItems,
+      msg.weaponsOnly ? (i) => i.bucket?.sort === 'Weapons' : () => true
+    );
+    loadout && (await dispatch(applyLoadout(store, loadout, { allowUndo: true })));
+  };
+}
+
+function collectPostmasterHandler({ store }: HandlerArgs<CollectPostmasterAction>): ThunkResult {
+  return async (dispatch) => dispatch(pullFromPostmaster(store));
+}
+
+// TODO move to a shared module
+function maximizePowerHandler({ state, store }: HandlerArgs<MaxPowerAction>): ThunkResult {
+  return async (dispatch) => {
+    const allItems = allItemsSelector(state);
+    const loadout = maxLightLoadout(allItems, store);
+    return dispatch(applyLoadout(store, loadout, { allowUndo: true }));
+  };
+}
+
+function farmingModeHandler({ state, store }: HandlerArgs<FarmingModeAction>): ThunkResult {
+  return async (dispatch) => {
+    if (state.farming.storeId) {
+      return dispatch(stopFarming());
+    } else {
+      return dispatch(startFarming(store?.id));
+    }
+  };
+}
+
+function selectionHandler({ msg, state }: HandlerArgs<SelectionAction>): ThunkResult {
+  return async (dispatch) => {
+    const selectionType = msg.selection;
+    dispatch(setSearchQuery(''));
+    dispatch(streamDeckWaitSelection(selectionType));
+    // open the related page
+    const path = selectionType === 'loadout' ? 'loadouts' : 'inventory';
+    if (!window.location.pathname.endsWith(path)) {
+      dispatch(setRouterLocation(routeTo(state, path)));
+      // delay a bit the notification show
+      await delay(100);
+    }
+    // show the notification
+    showSelectionNotification(state, selectionType, () => dispatch(streamDeckClearSelection()));
+  };
+}
+
+function equipLoadoutHandler({ msg, state }: HandlerArgs<EquipOrShareLoadoutAction>): ThunkResult {
+  return async (dispatch) => {
+    const loadouts = loadoutsSelector(state);
+    const stores = storesSelector(state);
+    const store = getStore(stores, msg.character);
+    const loadout = loadouts.find((it) => it.id === msg.loadout);
+    if (store && loadout) {
+      return dispatch(applyLoadout(store, loadout, { allowUndo: true }));
+    }
+  };
+}
+
+function shareLoadoutHandler({ msg, state }: HandlerArgs<EquipOrShareLoadoutAction>): ThunkResult {
+  return async (dispatch) => {
+    const loadouts = loadoutsSelector(state);
+    const account = currentAccountSelector(state);
+    const accountId = account?.membershipId;
+    const loadout = loadouts.find((it) => it.id === msg.loadout);
+    if (accountId && loadout) {
+      const shareUrl =
+        loadoutShares.get(loadout) ||
+        (await createLoadoutShare(accountId, convertDimLoadoutToApiLoadout(loadout)));
+      loadoutShares.set(loadout, shareUrl);
+      return dispatch(
+        sendToStreamDeck({
+          action: 'dim:share-url',
+          data: { shareUrl },
+        })
+      );
+    }
+  };
+}
+
+function freeBucketSlotHandler({
```

Comment:
This one's a neat action, but I'd like to posit a design constraint that the StreamDeck plugin may only trigger actions (or display information) that is already accessible within DIM. If we want to have a new type of operation like this, we need to make it possible without the StreamDeck *first*, then we can expose it to StreamDeck.

Reply:
ok that's not a problem, i can remove this feature; This idea was proposed by an user on my repo and I liked it since it's a manual/light version of farming mode, just a little more specific. How we can add this type of feature on DIM ? I mean I see this action as a random pick and move of an item to make space.

Reply:
So far we've advocated for people using searches or farming mode to achieve this sort of thing. It really depends on what the intended goal of the feature is - probably best to fork off into a separate discussion.

Reply:
ok